### ISSUE TITLE
setenv() fails sometimes

### ISSUE THREAD

**Issue by ingenieroariel** [jart/cosmopolitan#63]
cosmopolitan.nix
```
with import <nixpkgs> {};

stdenv.mkDerivation rec {
  name = "cosmopolitan-${version}";
  version = "0.2";
  src = fetchTarball {
    url = "https://justine.lol/cosmopolitan/cosmopolitan-0.2.tar.gz";
  };

  buildInputs = [ gnumake coreutils ctags gnused clang ];

  buildPhase = "${gnumake}/bin/make";

  MKDIR = "${coreutils}/bin/mkdir -p";

  installPhase = ''
    mkdir -p $out/lib
    cp *.o $out/lib
    cp *.a $out/lib
    cp *.lds $out/lib
    cp *.h $out/lib
  '';
}
```

Got error:

```
inymath.a o//libc/str/str.a o//libc/intrin/intrin.a o//libc/sysv/sysv.a o//libc/nt/kernel32.a o//libc/nexgen32e/nexgen32e.a o//libc/stubs/stubs.a -o o//test/libc/time/strftime_test.com.dbg
o/third_party/gcc/bin/x86_64-linux-musl-objcopy -S -O binary o//test/libc/time/strftime_test.com.dbg o//test/libc/time/strftime_test.com
o//test/libc/time/clock_gettime_test.com
o//test/libc/time/dsleep_test.com
Segmentation fault (core dumped)
o//test/libc/time/strftime_test.com
make: *** [build/rules.mk:76: o//test/libc/time/strftime_test.com.runs] Error 139
error: --- Error ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- nix-build
builder for '/nix/store/sffmakcsv51qws05gih80s085wwqd1md-cosmopolitan-0.2.drv' failed with exit code 2; last 10 log lines:
  o/third_party/gcc/bin/x86_64-linux-musl-objcopy -S -O binary o//test/libc/time/clock_gettime_test.com.dbg o//test/libc/time/clock_gettime_test.com
  o/third_party/gcc/bin/x86_64-linux-musl-ld.bfd -static -nostdlib -m elf_x86_64 --build-id=none --no-dynamic-linker -z max-page-size=0x1000 -T o//ape/ape.lds o//ape/ape.o o//libc/crt/crt.o o//libc/testlib/testmain.o o//third_party/getopt/getopt.a o//libc/testlib/runner.a o//test/libc/time/dsleep_test.o o//libc/testlib/testlib.a o//libc/x/x.a o//third_party/gdtoa/gdtoa.a o//libc/log/log.a o//libc/unicode/unicode.a o//libc/time/time.a o//libc/zipos/zipos.a o//third_party/zlib/zlib.a o//ape/lib/apelib.a o//libc/stdio/stdio.a o//libc/alg/alg.a o//libc/mem/mem.a o//third_party/dlmalloc/dlmalloc.a o//libc/bits/bits.a o//libc/runtime/runtime.a o//libc/elf/elf.a o//libc/rand/rand.a o//libc/calls/syscalls.a o//libc/sysv/calls.a o//libc/nt/ntdll.a o//libc/nt/advapi32.a o//libc/fmt/fmt.a o//third_party/compiler_rt/compiler_rt.a o//libc/tinymath/tinymath.a o//libc/str/str.a o//libc/intrin/intrin.a o//libc/sysv/sysv.a o//libc/nt/kernel32.a o//libc/nexgen32e/nexgen32e.a o//libc/stubs/stubs.a -o o//test/libc/time/dsleep_test.com.dbg
  o/third_party/gcc/bin/x86_64-linux-musl-objcopy -S -O binary o//test/libc/time/dsleep_test.com.dbg o//test/libc/time/dsleep_test.com
  o/third_party/gcc/bin/x86_64-linux-musl-ld.bfd -static -nostdlib -m elf_x86_64 --build-id=none --no-dynamic-linker -z max-page-size=0x1000 -T o//ape/ape.lds o//ape/ape.o o//libc/crt/crt.o o//libc/testlib/testmain.o o//third_party/getopt/getopt.a o//libc/testlib/runner.a o//test/libc/time/strftime_test.o o//libc/testlib/testlib.a o//libc/x/x.a o//third_party/gdtoa/gdtoa.a o//libc/log/log.a o//libc/unicode/unicode.a o//libc/time/time.a o//libc/zipos/zipos.a o//third_party/zlib/zlib.a o//ape/lib/apelib.a o//libc/stdio/stdio.a o//libc/alg/alg.a o//libc/mem/mem.a o//third_party/dlmalloc/dlmalloc.a o//libc/bits/bits.a o//libc/runtime/runtime.a o//libc/elf/elf.a o//libc/rand/rand.a o//libc/calls/syscalls.a o//libc/sysv/calls.a o//libc/nt/ntdll.a o//libc/nt/advapi32.a o//libc/fmt/fmt.a o//third_party/compiler_rt/compiler_rt.a o//libc/tinymath/tinymath.a o//libc/str/str.a o//libc/intrin/intrin.a o//libc/sysv/sysv.a o//libc/nt/kernel32.a o//libc/nexgen32e/nexgen32e.a o//libc/stubs/stubs.a -o o//test/libc/time/strftime_test.com.dbg
  o/third_party/gcc/bin/x86_64-linux-musl-objcopy -S -O binary o//test/libc/time/strftime_test.com.dbg o//test/libc/time/strftime_test.com
  o//test/libc/time/clock_gettime_test.com
  o//test/libc/time/dsleep_test.com
  Segmentation fault (core dumped)
  o//test/libc/time/strftime_test.com
  make: *** [build/rules.mk:76: o//test/libc/time/strftime_test.com.runs] Error 139
```

the way to run the file I shared is `nix-build cosmopolitan.nix` after installing nix with `curl -L https://nixos.org/nix/install | sh`

**Comment by lourkeur** [comment]
Can reproduce the same failure on NixOS.

My [attempt](https://github.com/lourkeur/nixpkgs/blob/cosmopolitan/pkgs/development/compilers/cosmopolitan/default.nix) from last week didn't get as far:
```
o/third_party/gcc/bin/x86_64-linux-musl-gcc -msse3 -mno-red-zone -fno-math-errno -fno-trapping-math -fno-fp-int-builtin-inexact -fno-ident -fno-common -fno-gnu-unique -fstrict-aliasing -fstrict-overflow -fno-semantic-interposition -mno-omit-leaf-frame-pointer -Wall -Werror -fdebug-prefix-map=/build/source= -frecord>
  o/third_party/gcc/bin/x86_64-linux-musl-gcc -msse3 -mno-red-zone -fno-math-errno -fno-trapping-math -fno-fp-int-builtin-inexact -fno-ident -fno-common -fno-gnu-unique -fstrict-aliasing -fstrict-overflow -fno-semantic-interposition -mno-omit-leaf-frame-pointer -Wall -Werror -fdebug-prefix-map=/build/source= -frecord>
  o/third_party/gcc/bin/x86_64-linux-musl-gcc -msse3 -mno-red-zone -fno-math-errno -fno-trapping-math -fno-fp-int-builtin-inexact -fno-ident -fno-common -fno-gnu-unique -fstrict-aliasing -fstrict-overflow -fno-semantic-interposition -mno-omit-leaf-frame-pointer -Wall -Werror -fdebug-prefix-map=/build/source= -frecord>
  o/third_party/gcc/bin/x86_64-linux-musl-gcc -msse3 -mno-red-zone -fno-math-errno -fno-trapping-math -fno-fp-int-builtin-inexact -fno-ident -fno-common -fno-gnu-unique -fstrict-aliasing -fstrict-overflow -fno-semantic-interposition -mno-omit-leaf-frame-pointer -Wall -Werror -fdebug-prefix-map=/build/source= -frecord>
  o/third_party/gcc/bin/x86_64-linux-musl-gcc -msse3 -mno-red-zone -fno-math-errno -fno-trapping-math -fno-fp-int-builtin-inexact -fno-ident -fno-common -fno-gnu-unique -fstrict-aliasing -fstrict-overflow -fno-semantic-interposition -mno-omit-leaf-frame-pointer -Wall -Werror -fdebug-prefix-map=/build/source= -frecord>
  o/third_party/gcc/bin/x86_64-linux-musl-gcc -msse3 -mno-red-zone -fno-math-errno -fno-trapping-math -fno-fp-int-builtin-inexact -fno-ident -fno-common -fno-gnu-unique -fstrict-aliasing -fstrict-overflow -fno-semantic-interposition -mno-omit-leaf-frame-pointer -Wall -Werror -fdebug-prefix-map=/build/source= -frecord>
  o/third_party/gcc/bin/x86_64-linux-musl-gcc -msse3 -mno-red-zone -fno-math-errno -fno-trapping-math -fno-fp-int-builtin-inexact -fno-ident -fno-common -fno-gnu-unique -fstrict-aliasing -fstrict-overflow -fno-semantic-interposition -mno-omit-leaf-frame-pointer -Wall -Werror -fdebug-prefix-map=/build/source= -frecord>
  o/third_party/gcc/bin/x86_64-linux-musl-gcc -msse3 -mno-red-zone -fno-math-errno -fno-trapping-math -fno-fp-int-builtin-inexact -fno-ident -fno-common -fno-gnu-unique -fstrict-aliasing -fstrict-overflow -fno-semantic-interposition -mno-omit-leaf-frame-pointer -Wall -Werror -fdebug-prefix-map=/build/source= -frecord>
  make: sh: Argument list too long
  make: *** [build/rules.mk:39: o//cosmopolitan.a] Error 127
```

**Comment by lourkeur** [comment]
It will build successfully under `nix-shell cosmopolitan.nix` so it's something between the Nix sandbox and one of the strftime tests...

**Comment by lourkeur** [comment]
It's segfaulting in setenv. If you patch it out it should build

```
with import <nixpkgs> {};

stdenv.mkDerivation rec {
  name = "cosmopolitan-${version}";
  version = "0.2";
  src = fetchTarball {
    url = "https://justine.lol/cosmopolitan/cosmopolitan-0.2.tar.gz";
  };

  buildInputs = [ gnumake coreutils ctags gnused clang ];

  buildPhase = "${gnumake}/bin/make";

  MKDIR = "${coreutils}/bin/mkdir -p";

  patches = [ (writeText "skip-setenv.patch" ''
--- a/test/libc/time/strftime_test.c
+++ b/test/libc/time/strftime_test.c
@@ -23,7 +23,6 @@
 #include "libc/time/time.h"
 
 textstartup static void strftime_test_init(void) {
-  setenv("TZ", "GST", true);
 }
 const void *const strftime_test_ctor[] initarray = {strftime_test_init};
 
@@ -57,24 +56,6 @@ TEST(strftime_100, rfc822_ShakaZuluTime) {
                FormatTime("%a, %d %b %y %T %z", gmtime(&t)));
 }
 
-TEST(strftime_201, iso8601_GoogleStandardTime) {
-  int64_t t = 0x5cd04d0e;
-  ASSERT_STREQ("2019-05-06T08:04:46PDT",
-               FormatTime("%Y-%m-%dT%H:%M:%S%Z", localtime(&t)));
-}
-
-TEST(strftime_201, rfc2822_GoogleStandardTime) {
-  int64_t t = 0x5cd04d0e;
-  ASSERT_STREQ("Mon, 06 May 2019 08:04:46 -0700",
-               FormatTime("%a, %d %b %Y %T %z", localtime(&t)));
-}
-
-TEST(strftime_201, rfc822_GoogleStandardTime) {
-  int64_t t = 0x5cd04d0e;
-  ASSERT_STREQ("Mon, 06 May 19 08:04:46 -0700",
-               FormatTime("%a, %d %b %y %T %z", localtime(&t)));
-}
-
 /* TEST(xiso8601, testModernity_TODO) { */
 /*   int64_t t = (1600 - 1970) * 31536000; */
 /*   ASSERT_STREQ("1600-01-01T00:00:00Z", */
'')];

  installPhase = ''
    mkdir -p $out/lib
    cp *.o $out/lib
    cp *.a $out/lib
    cp *.lds $out/lib
    cp *.h $out/lib
  '';
}
```

This isn't a long term solution but at least we figured that out.

**Comment by ingenieroariel** [comment]
Thanks Louis!

**Comment by jart** [comment]
Thanks for the super helpful triage @lourkeur. I noticed this same issue several days ago. It was basically an initialization order issue. I fixed it in cdc54ea1fd86253ced49dd24d45660cadac1d018. I see that you're using the 0.2 release. You can download cosmopolitan.zip at https://justine.lol/cosmopolitan/download.html to get the fixed nightly release. I intend to roll a 0.3 release very soon.
